File: scrapers/browser.js
// Stealth browser with human-like behavior
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');

puppeteer.use(StealthPlugin());

// ── Human-like delays ───────────────────────────────────────────────────────
function randomDelay(min = 800, max = 2500) {
  const delay = Math.floor(Math.random() * (max - min + 1)) + min;
  return new Promise(resolve => setTimeout(resolve, delay));
}

function typingDelay() {
  return Math.floor(Math.random() * 150) + 50; // 50-200ms between keystrokes
}

// ── React-friendly typing ───────────────────────────────────────────────────
// React controlled inputs need native input events to update state.
// Standard puppeteer keyboard.type dispatches keydown/keypress/keyup but
// React listens on the native input event via its synthetic event system.
async function reactType(page, selector, text) {
  await page.waitForSelector(selector, { timeout: 15000 });

  // Click to focus
  const el = await page.$(selector);
  await el.click({ clickCount: 3 }); // select all existing text
  await randomDelay(200, 400);
  await page.keyboard.press('Backspace'); // clear
  await randomDelay(200, 400);

  // Type each character and trigger React-compatible events
  for (const char of text) {
    await page.keyboard.type(char, { delay: typingDelay() });
  }

  // Force React to pick up the value by dispatching native events
  await page.evaluate((sel, val) => {
    const input = document.querySelector(sel);
    if (input) {
      // Use React's internal value setter to bypass controlled component
      const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
        window.HTMLInputElement.prototype, 'value'
      ).set;
      nativeInputValueSetter.call(input, val);

      // Dispatch events React listens to
      input.dispatchEvent(new Event('input', { bubbles: true }));
      input.dispatchEvent(new Event('change', { bubbles: true }));
    }
  }, selector, text);

  await randomDelay(300, 600);
}

// ── Human-like typing (non-React) ───────────────────────────────────────────
async function humanType(page, selector, text) {
  await page.waitForSelector(selector, { timeout: 15000 });
  await page.click(selector);
  await randomDelay(300, 700);

  for (const char of text) {
    await page.keyboard.type(char, { delay: typingDelay() });
  }
  await randomDelay(200, 500);
}

// ── Human-like click ────────────────────────────────────────────────────────
async function humanClick(page, selector) {
  await page.waitForSelector(selector, { timeout: 15000 });

  const element = await page.$(selector);
  const box = await element.boundingBox();

  if (box) {
    const x = box.x + box.width / 2 + (Math.random() * 6 - 3);
    const y = box.y + box.height / 2 + (Math.random() * 6 - 3);

    await page.mouse.move(x, y, { steps: Math.floor(Math.random() * 10) + 5 });
    await randomDelay(100, 300);
    await page.mouse.click(x, y);
  } else {
    await page.click(selector);
  }

  await randomDelay(500, 1500);
}

// ── Random scroll to simulate reading ───────────────────────────────────────
async function humanScroll(page, amount = null) {
  const scrollAmount = amount || Math.floor(Math.random() * 400) + 200;
  await page.evaluate((px) => {
    window.scrollBy({ top: px, behavior: 'smooth' });
  }, scrollAmount);
  await randomDelay(500, 1500);
}

// ── Launch stealth browser ──────────────────────────────────────────────────
async function launchBrowser() {
  const browser = await puppeteer.launch({
    headless: process.env.CI === 'true' ? 'new' : false,
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-blink-features=AutomationControlled',
      '--disable-features=IsolateOrigins,site-per-process',
      '--disable-web-security',
      '--disable-features=CrossSiteDocumentBlockingIfIsolating',
      '--window-size=1920,1080',
      '--lang=en-US,en',
    ],
    defaultViewport: { width: 1920, height: 1080 },
    ignoreDefaultArgs: ['--enable-automation'],
  });

  return browser;
}

// ── Create page with stealth settings ───────────────────────────────────────
async function createPage(browser) {
  const page = await browser.newPage();

  // Set realistic user agent
  await page.setUserAgent(
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36'
  );

  await page.setViewport({ width: 1920, height: 1080 });

  // Extra stealth: override webdriver, plugins, permissions
  await page.evaluateOnNewDocument(() => {
    // Hide webdriver
    Object.defineProperty(navigator, 'webdriver', { get: () => false });

    // Language
    Object.defineProperty(navigator, 'language', { get: () => 'en-US' });
    Object.defineProperty(navigator, 'languages', { get: () => ['en-US', 'en'] });
    Object.defineProperty(navigator, 'platform', { get: () => 'MacIntel' });

    // Fake plugins
    Object.defineProperty(navigator, 'plugins', {
      get: () => [
        { name: 'Chrome PDF Plugin', filename: 'internal-pdf-viewer' },
        { name: 'Chrome PDF Viewer', filename: 'mhjfbmdgcfjbbpaeojofohoefgiehjai' },
        { name: 'Native Client', filename: 'internal-nacl-plugin' },
      ],
    });

    // Fake permissions
    const originalQuery = window.navigator.permissions.query;
    window.navigator.permissions.query = (parameters) =>
      parameters.name === 'notifications'
        ? Promise.resolve({ state: Notification.permission })
        : originalQuery(parameters);

    // Chrome runtime
    window.chrome = { runtime: {}, loadTimes: () => {}, csi: () => {} };

    // WebGL vendor
    const getParameter = WebGLRenderingContext.prototype.getParameter;
    WebGLRenderingContext.prototype.getParameter = function(parameter) {
      if (parameter === 37445) return 'Intel Inc.';
      if (parameter === 37446) return 'Intel Iris OpenGL Engine';
      return getParameter.call(this, parameter);
    };
  });

  return page;
}

// ── Navigate with human-like behavior ───────────────────────────────────────
async function humanNavigate(page, url, options = {}) {
  await page.goto(url, {
    waitUntil: 'networkidle2',
    timeout: 30000,
    ...options,
  });
  await randomDelay(1500, 3500);
  await humanScroll(page, Math.floor(Math.random() * 200) + 50);
}

module.exports = {
  launchBrowser,
  createPage,
  humanType,
  reactType,
  humanClick,
  humanScroll,
  humanNavigate,
  randomDelay,
};

File: scrapers/browser.py
"""Camoufox browser + human-like helpers.

Camoufox is a Firefox build with anti-fingerprinting patched into the binary,
so we don't need stealth plugins. We use a persistent profile so cookies and
sessions survive between runs.
"""

from __future__ import annotations

import os
import random
import time
from contextlib import contextmanager
from pathlib import Path
from typing import Iterator

from camoufox.sync_api import Camoufox
from playwright.sync_api import BrowserContext, Page

PROFILE_DIR = Path(__file__).resolve().parent.parent / "firefox-profile"
DEBUG_DIR = Path(__file__).resolve().parent.parent / "debug"


def random_delay(min_ms: int = 800, max_ms: int = 2500) -> None:
    time.sleep(random.uniform(min_ms / 1000, max_ms / 1000))


@contextmanager
def launch_browser(headless: bool = False) -> Iterator[BrowserContext]:
    """Yield a camoufox persistent BrowserContext (cookies survive between runs)."""
    PROFILE_DIR.mkdir(parents=True, exist_ok=True)
    DEBUG_DIR.mkdir(parents=True, exist_ok=True)

    # CI=true → run headless; otherwise show the window so you can solve any
    # one-off CAPTCHA / 2FA prompts manually.
    if os.environ.get("CI") == "true":
        headless = True

    with Camoufox(
        headless=headless,
        humanize=True,           # mouse moves along Bezier curves
        geoip=True,              # spoofed locale matches the IP's geo
        locale="en-US",
        os=("macos",),
        persistent_context=True,
        user_data_dir=str(PROFILE_DIR),
        window=(1440, 900),
    ) as browser:
        yield browser


def new_page(ctx: BrowserContext) -> Page:
    """Reuse the persistent context's existing tab if it has one, else open a new one."""
    return ctx.pages[0] if ctx.pages else ctx.new_page()


# ── Human-like input ────────────────────────────────────────────────────────
def human_type(page: Page, selector: str, text: str) -> None:
    """Click the field, clear it, type char-by-char with jittered delays."""
    page.wait_for_selector(selector, timeout=15000)
    page.click(selector, click_count=3)  # select existing
    random_delay(200, 400)
    page.keyboard.press("Backspace")
    random_delay(200, 400)
    for ch in text:
        page.keyboard.type(ch, delay=random.randint(60, 180))
    random_delay(200, 500)


def react_type(page: Page, selector: str, text: str) -> None:
    """For React controlled inputs that ignore plain keyboard.type events."""
    human_type(page, selector, text)
    page.evaluate(
        """([sel, val]) => {
            const el = document.querySelector(sel);
            if (!el) return;
            const setter = Object.getOwnPropertyDescriptor(
                window.HTMLInputElement.prototype, 'value'
            ).set;
            setter.call(el, val);
            el.dispatchEvent(new Event('input', { bubbles: true }));
            el.dispatchEvent(new Event('change', { bubbles: true }));
        }""",
        [selector, text],
    )
    random_delay(300, 600)


def human_click(page: Page, selector: str) -> None:
    """Move to the element with a curve-ish path, then click. Camoufox's
    humanize=True already smooths real mouse moves; this just jitters the
    landing point inside the bounding box."""
    page.wait_for_selector(selector, timeout=15000)
    el = page.query_selector(selector)
    box = el.bounding_box() if el else None
    if box:
        x = box["x"] + box["width"] / 2 + random.uniform(-3, 3)
        y = box["y"] + box["height"] / 2 + random.uniform(-3, 3)
        page.mouse.move(x, y)
        random_delay(100, 300)
        page.mouse.click(x, y)
    else:
        page.click(selector)
    random_delay(500, 1500)


def human_scroll(page: Page, amount: int | None = None) -> None:
    px = amount if amount is not None else random.randint(200, 600)
    page.evaluate("(px) => window.scrollBy({top: px, behavior: 'smooth'})", px)
    random_delay(500, 1500)


def human_navigate(page: Page, url: str, **goto_kwargs) -> None:
    page.goto(url, wait_until="domcontentloaded", timeout=30000, **goto_kwargs)
    random_delay(1500, 3500)
    human_scroll(page, random.randint(50, 250))


# ── Debug snapshot ──────────────────────────────────────────────────────────
def debug_snapshot(page: Page, label: str) -> None:
    """Save fullpage screenshot + HTML to debug/ for failure forensics."""
    try:
        png = DEBUG_DIR / f"{label}.png"
        html = DEBUG_DIR / f"{label}.html"
        page.screenshot(path=str(png), full_page=True)
        html.write_text(page.content(), encoding="utf-8")
        print(f"  📸 Debug saved: debug/{label}.png + .html")
    except Exception:
        pass
